src/support/bits.h

```diff
@@ -66,14 +66,14 @@ int CountLeadingZeroes(T v) {
   return CountLeadingZeroes(typename std::make_unsigned<T>::type(v));
 }
 
-template <typename T>
-inline static T RotateLeft(T val, T count) {
+template <typename T, typename U>
```

Comment:
Thanks for the detailed answers. I now see that binary encodings of dyadic operators contain no operand typing and therefore both operands must be the same type as the opcode. Lot's to learn. 

    On Thursday, May 12, 2016 1:20 AM, Derek Schuff <notifications@github.com> wrote:
 

 In src/support/bits.h:> @@ -66,14 +66,14 @@ int CountLeadingZeroes(T v) {
>    return CountLeadingZeroes(typename std::make_unsigned<T>::type(v));
>  }
>  
> -template <typename T>
> -inline static T RotateLeft(T val, T count) {
> +template <typename T, typename U>
The WebAssembly semantics are that both operands are the same size, but that the shift count bits beyond the 5 least significant are ignored. C's semantics are that if the shift count is too large (e.g. if any of those bits are set) then the behavior is undefined (that's a very bad thing). The mask operation in the first 2 lines of the implementation is designed to implement the wasm rotate semantics but if it fails to mask all the higher-order bits (which could happen if U were accidentally larger than T) then the behavior of the interpreter would be undefined according to C shift semantics.The effect of -count is, bitwise, the same if count is a signed or unsigned int, so -5 is 0xFFFFFFFB. and the right-shift count becomes 0xB (or 11, which is ) after masking. Other than that, this particular implementation only does bitwise operations and so is agnostic to whether the types are signed or unsigned.—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub